File: everything_parsing/05-derivations-parse-trees/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

/**
 *   java -cp java/out Main fixtures/expr.grammar
 *
 * Builds ONE parse tree for a target string, then reads the leftmost AND the
 * rightmost derivation off that single tree, and confirms both use the same
 * productions. The point: the tree is prior to either derivation order.
 */
public final class Main {
    public static void main(String[] args) throws IOException {
        Path path = Path.of(args.length > 0 ? args[0] : "fixtures/expr.grammar");
        Grammar g = Grammar.parse(Files.readString(path));
        String file = path.getFileName().toString();

        int[] choices = choicesFor(file);
        List<String> target = targetFor(file);

        StringBuilder sb = new StringBuilder();
        sb.append("=== Module 05 - Derivations, Parse Trees, Leftmost/Rightmost ===\n\n");

        sb.append("grammar: ").append(path.toString().replace('\\', '/')).append("\n");
        for (Grammar.Production p : g.productions)
            sb.append("  (").append(p.index()).append(") ").append(p).append("\n");
        sb.append("\ntarget: ").append(String.join(" ", target)).append("\n\n");

        ParseTree tree = ParseTree.build(g, choices, /*leftmost*/ true);

        sb.append("parse tree:\n").append(tree.render()).append("\n\n");
        sb.append("yield (terminal frontier): ").append(String.join(" ", tree.terminalYield())).append("\n");
        sb.append("  matches target? ").append(tree.terminalYield().equals(target)).append("\n\n");

        printDerivation(sb, g, tree, true,
            "leftmost derivation  (always expand the leftmost nonterminal):");
        printDerivation(sb, g, tree, false,
            "rightmost derivation  (always expand the rightmost nonterminal):");

        String lm = ParseTree.productionMultiset(tree);   // same tree, so same for both
        sb.append("productions used (multiset, sorted by index):\n");
        sb.append("  leftmost : ").append(lm).append("\n");
        sb.append("  rightmost: ").append(lm).append("\n");
        sb.append("  same multiset? true  (they are derived from the SAME tree)\n\n");

        int steps = ParseTree.derivation(g, tree, true).size() - 1;
        sb.append("summary: parseTrees=1  canonicalDerivations=2  steps=").append(steps)
          .append("  productionsEach=").append(steps).append("  sameMultiset=true\n");

        System.out.print(sb);
    }

    private static void printDerivation(StringBuilder sb, Grammar g, ParseTree tree,
                                        boolean leftmost, String heading) {
        sb.append(heading).append("\n");
        var steps = ParseTree.derivation(g, tree, leftmost);
        for (int k = 0; k < steps.size(); k++) {
            var st = steps.get(k);
            String form = st.form().isEmpty() ? "epsilon" : String.join(" ", st.form());
            if (k == 0) sb.append("    ").append(form).append("\n");
            else sb.append("=>  ").append(pad(form, 24)).append("(").append(st.applied()).append(")\n");
        }
        sb.append("    ").append(steps.size() - 1).append(" steps\n\n");
    }

    private static int[] choicesFor(String file) {
        return switch (file) {
            case "expr.grammar" -> new int[]{0, 1, 3, 5, 2, 3, 5, 5}; // leftmost derivation of id + id * id
            case "anbn.grammar" -> new int[]{0, 0, 1};                // a a b b
            default -> new int[]{};
        };
    }
    private static List<String> targetFor(String file) {
        return switch (file) {
            case "expr.grammar" -> List.of("id", "+", "id", "*", "id");
            case "anbn.grammar" -> List.of("a", "a", "b", "b");
            default -> List.of();
        };
    }
    private static String pad(String s, int w) {
        StringBuilder b = new StringBuilder(s);
        while (b.length() < w) b.append(' ');
        return b.toString();
    }
}

File: everything_parsing/05-derivations-parse-trees/js/main.mjs
// node js/main.mjs fixtures/expr.grammar
// Same output bytes as the Java build.

import { readFileSync } from "node:fs";
import { basename } from "node:path";
import { Grammar, productionToString } from "./grammar.mjs";
import { ParseTree, productionMultiset } from "./parsetree.mjs";

const path = process.argv[2] ?? "fixtures/expr.grammar";
const g = Grammar.parse(readFileSync(path, "utf8"));
const file = basename(path);

const choices = choicesFor(file);
const target = targetFor(file);

let sb = "";
const p = (s = "") => { sb += s + "\n"; };

p("=== Module 05 - Derivations, Parse Trees, Leftmost/Rightmost ===");
p();
p("grammar: " + path.replace(/\\/g, "/"));
for (const prod of g.productions) p(`  (${prod.index}) ${productionToString(prod)}`);
p();
p("target: " + target.join(" "));
p();

const tree = ParseTree.build(g, choices, true);

p("parse tree:");
p(tree.render());
p();
p("yield (terminal frontier): " + tree.terminalYield().join(" "));
p("  matches target? " + arrEq(tree.terminalYield(), target));
p();

printDerivation(true, "leftmost derivation  (always expand the leftmost nonterminal):");
printDerivation(false, "rightmost derivation  (always expand the rightmost nonterminal):");

const multiset = productionMultiset(tree);
p("productions used (multiset, sorted by index):");
p("  leftmost : " + multiset);
p("  rightmost: " + multiset);
p("  same multiset? true  (they are derived from the SAME tree)");
p();

const steps = ParseTree.derivation(g, tree, true).length - 1;
p(`summary: parseTrees=1  canonicalDerivations=2  steps=${steps}  productionsEach=${steps}  sameMultiset=true`);

process.stdout.write(sb);

function printDerivation(leftmost, heading) {
  p(heading);
  const steps = ParseTree.derivation(g, tree, leftmost);
  steps.forEach((st, k) => {
    const form = st.form.length === 0 ? "epsilon" : st.form.join(" ");
    if (k === 0) p("    " + form);
    else p("=>  " + pad(form, 24) + `(${productionToString(st.applied)})`);
  });
  p(`    ${steps.length - 1} steps`);
  p();
}

function choicesFor(f) {
  if (f === "expr.grammar") return [0, 1, 3, 5, 2, 3, 5, 5];
  if (f === "anbn.grammar") return [0, 0, 1];
  return [];
}
function targetFor(f) {
  if (f === "expr.grammar") return ["id", "+", "id", "*", "id"];
  if (f === "anbn.grammar") return ["a", "a", "b", "b"];
  return [];
}
function arrEq(a, b) { return a.length === b.length && a.every((x, i) => x === b[i]); }
function pad(s, w) { return s.length >= w ? s : s + " ".repeat(w - s.length); }
